libs/ui-components/src/model-viewer/components/model-upload/model-upload.component.ts

```
@@ -0,0 +1,234 @@
+        ),
+      )
+      .subscribe((result) => this.handleConfigUploadResponse(result));
+  }
+
+  public handleConfigUploadResponse(_result: unknown) {
+    const result = _result as UploadProgress | ModelConfigurationEntity;
+    if ('material' in result) {
+      this.resetUploadState();
+      this.updateConfigSnapshot(result);
+    }
+  }
+
+  public saveActiveConfig(id: string) {
+    this.localStorageService.setLocalItem(LocalStorageKeys.ACTIVE_CONFIG, id);
+  }
+
+  public async updateConfigSnapshot({ id, file }: ModelConfigurationEntity) {
+    this.saveActiveConfig(id);
+    const modelUrl = file.url;
+
+    this.store.dispatch(ModelListActions.updateModelSnapshotStart());
+
+    const snapshot = await this.handleSnapshotGeneration(modelUrl);
+    this.uploadService
+      .updateConfigSnapshot(snapshot, this.handleFailedConfigSnapshotUpdate)
+      .subscribe({
+        next: async (updatedConfig) => {
+          this.store.dispatch(
+            ModelListActions.addModel({
+              model: updatedConfig as ModelConfigurationEntity,
+            }),
+          );
+
+          this.store.dispatch(ModelListActions.updateModelSnapshotComplete());
+
+          await this.router.navigate([NAVIGATION_ROUTES.MOLDING_CONFIGURATION]);
+        },
+        error: () => {
+          this.store.dispatch(
+            ModelListActions.updateModelSnapshotFailure({
+              id,
+              error: 'Failed to update snapshot',
+            }),
+          );
+          this.handleFailedConfigSnapshotUpdate();
+        },
+      });
+  }
+
+  public handleFailedConfigSnapshotUpdate() {
+    this.showSnackbar('Failed to update config snapshot', 'error-snackbar');
+  }
+
+  public async handleSnapshotGeneration(modelUrl: string): Promise<string> {
+    this.uploadedFileUrl.set(modelUrl);
+    return (await this.modelViewerService.getCanvasOutput(
+      modelUrl,
+      true,
+      undefined,
+      undefined,
+      1.5,
```

This is ambiguous, if it is a constant kindly make it so. 
but you and casting is like 1 and 2. I wonder how you are handling this in the UI. 
oops! that's a private method right?

---

Now I see, imagine `handleSnapshotGeneration` returns an empty string or say ` 'undefined'` what happens to the consumer, It takes that and fails silently and I bet you may not be able to trace these kinds of errors. 
So this kind of casting is very dangerous, avoid it.

---

Alright